### ISSUE TITLE
Translation using composed keys from multiple source not merge correctly

### ISSUE THREAD

**Issue by FabienDehopre** [robisim74/angular-l10n#152]
**I'm submitting a...**
```
[x] bug report
[ ] feature request
[ ] support request
```

**Expected behavior**
The translation object should be merged recursively.

**Actual behavior**
For the moment, when the translations are loaded, only the first level are merged correctly.

**Steps to reproduce the behavior**
Here is an example of translations sources and the requested result when merged recursively:
Source1:
```json
{
  "KERNEL": {
    "MSG": {
      "REQUIRED": "Ce champs est obligatoire.",
      "ERRORS_TITLE": "Erreur(s) bloquante(s)",
      "ERRORS_BADGE": "{{count}} erreur(s)",
      "WARNINGS_TITLE": "Alerte(s)",
      "WARNINGS_BADGE": "{{count}} alerte(s)",
      "INTERNAL_ERROR": "Une erreur s'est produite.",
      "DEFAULT_SUCCESS": "L'opération s'est déroulée correctement.",
      "DEFAULT_WARNING": "L'opération s'est déroulée avec des alertes.",
      "DEFAULT_ERROR": "Une erreur s'est produite."
    }
  }
}
```
Source2:
```json
{
  "SHARED": {
    "MSG": {
      "EMAIL_FORMAT": "L'adresse e-mail n'est pas valide.",
      "INVALID_IBAN": "L'IBAN n'est pas valide.",
      "BIC_FORMAT": "Le code BIC n'est pas valide."
    },
    "CALENDAR": {
      "TODAY": "Aujoudr'hui",
      "CLEAR": "Effacer"
    }
  }
}
```
Source3:
```json
{
  "MSG": "Message de test",
  "FORM": {
    "BANQUE": "Banque",
    "CONTACT": "Contact",
    "PERSONNE": "Personne",
    "IBAN": {
      "LABEL": "IBAN",
      "PLACEHOLDER": "Introduisez votre IBAN"
    },
    "BIC": {
      "LABEL": "BIC",
      "PLACEHOLDER": "Introduisez votre BIC"
    },
    "TELEPHONE": {
      "LABEL": "Téléphone",
      "PLACEHOLDER": "Introduisez votre n° de téléphone"
    },
    "GSM": {
      "LABEL": "GSM",
      "PLACEHOLDER": "Introduisez votre n° de GSM"
    },
    "FAX": {
      "LABEL": "Fax",
      "PLACEHOLDER": "Introduisez votre n° de fax"
    },
    "EMAIL": {
      "LABEL": "E-mail",
      "PLACEHOLDER": "Introduisez votre adresse e-mail"
    },
    "NOM": {
      "LABEL": "Nom",
      "PLACEHOLDER": "Introduisez votre nom"
    },
    "PRENOM": {
      "LABEL": "Prénom",
      "PLACEHOLDER": "Introduisez votre adresse prénom"
    },
    "NISS": {
      "LABEL": "N° National",
      "PLACEHOLDER": "Introduisez votre n° national"
    },
    "DATE_NAISSANCE": {
      "LABEL": "Date de naissance",
      "PLACEHOLDER": "Introduisez votre date de naissance"
    },
    "SEXE": {
      "LABEL": "Sexe",
      "PLACEHOLDER": "Sélectionnez votre sexe"
    },
    "LANGUE": {
      "LABEL": "Langue",
      "PLACEHOLDER": "Sélectionnez votre langue"
    }
  }
}
```
Source4: 
```json
{
  "KERNEL": {
    "ERROR": {
      "1": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "2": "La valeur du champ {{propertyName}} doit être inférieure à {{valueMin}}",
      "3": "La valeur du champ {{propertyName}} doit être inférieure ou égale à {{valueMin}}",
      "4": "La valeur du champ {{propertyName}} doit être supérieure à {{valueMin}}",
      "5": "La valeur du champ {{propertyName}} doit être supérieure ou égale à {{valueMin}}",
      "6": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "7": "La valeur du champ {{propertyName}} doit être inférieure à {{valueMin}}",
      "8": "La valeur du champ {{propertyName}} doit être inférieure ou égale à {{valueMin}}",
      "9": "La valeur du champ {{propertyName}} doit être supérieure à {{valueMin}}",
      "10": "La valeur du champ {{propertyName}} doit être supérieure ou égale à {{valueMin}}",
      "11": "La valeur du champ {{propertyName}} doit être entre {{valueMin}} et {{valueMax}}",
      "13": "La valeur du champ {{propertyName}} doit être numérique",
      "14": "La valeur du champ {{propertyName}} doit être égale à une de ces valeurs: {{valueMin}}, {{valueMax}}, ...",
      "30": "La valeur du champ {{propertyName}} doit être inférieure à {{valueMin}}",
      "31": "La valeur du champ {{propertyName}} doit être supérieure à {{valueMin}}",
      "32": "La valeur du champ {{propertyName}} doit être inférieure ou égale à {{valueMin}}",
      "33": "La valeur du champ {{propertyName}} doit être supérieure ou égale à {{valueMin}}",
      "34": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "35": "La valeur du champ {{propertyName}} doit être inclu entre {{valueMin}} et {{valueMax}}",
      "36": "La valeur du champ {{propertyName}} doit être inférieure à {{valueMin}}",
      "37": "La valeur du champ {{propertyName}} doit être supérieure à {{valueMin}}",
      "38": "La valeur du champ {{propertyName}} doit être inférieure ou égale à {{valueMin}}",
      "39": "La valeur du champ {{propertyName}} doit être supérieure ou égale à {{valueMin}}",
      "40": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "41": "La valeur du champ {{propertyName}} doit être égale à une de ces valeurs: {{valueMin}}, {{valueMax}}, ...",
      "60": "La valeur du champ {{propertyName}} ne peux pas être rempli",
      "61": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "62": "La valeur du champ {{propertyName}} doit être numérique",
      "63": "La longueur du champ {{propertyName}} n'est pas correcte ({{valueMin}})",
      "64": "La valeur du champ {{propertyName}} ne peux pas être rempli",
      "65": "La valeur du champ {{propertyName}} doit être égale à une de ces valeurs: {{valueMin}}, {{valueMax}}, ...",
      "66": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "67": "La valeur du champ {{propertyName}} doit avoir une longueur comprise entre {{valueMin}} et {{valueMax}}",
      "90": "Le champ {{propertyName}} est obligatoire",
      "91": "L'objet {{propertyName}} existe déjà dans l'application",
      "92": "L'objet {{propertyName}} n'a pas été trouvé",
      "93": "La valeur du champ {{propertyName}} est immutable et ne peux pas être modifié",
      "94": "La valeur du champ {{propertyName}} est obligatoire",
      "110": "Le nombre d'élements {{propertyName}} doit être compris entre {{valueMin}} et {{valueMax}}",
      "111": "Le nombre d'élements {{propertyName}} doit être compris entre {{valueMin}} et {{valueMax}}",
      "112": "La valeur du champ {{propertyName}} est en chevauchement avec {{valueMin}}",
      "113": "La valeur du champ {{propertyName}} ne se trouve pas dans la liste",
      "114": "L'objet {{propertyName}} n'est pas accessible par cet utilisateur",
      "130": "Le format du champ {{propertyName}} ne représente pas un numéro de téléphone valable",
      "131": "Le format du champ {{propertyName}} ne représente pas un e-mail valable",
      "150": "Le format du champ {{propertyName}} est incorrect",
      "199": "Un bug s'est produit"
    }
  },
  "EXTERNAL": {
    "ERROR": {
      "201": "Le service {{source}} est indisponible : {{message}}"
    }
  }
}
```
Actual result when using spread operator:
```json
{
  "KERNEL": {
    "ERROR": {
      "1": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "2": "La valeur du champ {{propertyName}} doit être inférieure à {{valueMin}}",
      "3": "La valeur du champ {{propertyName}} doit être inférieure ou égale à {{valueMin}}",
      "4": "La valeur du champ {{propertyName}} doit être supérieure à {{valueMin}}",
      "5": "La valeur du champ {{propertyName}} doit être supérieure ou égale à {{valueMin}}",
      "6": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "7": "La valeur du champ {{propertyName}} doit être inférieure à {{valueMin}}",
      "8": "La valeur du champ {{propertyName}} doit être inférieure ou égale à {{valueMin}}",
      "9": "La valeur du champ {{propertyName}} doit être supérieure à {{valueMin}}",
      "10": "La valeur du champ {{propertyName}} doit être supérieure ou égale à {{valueMin}}",
      "11": "La valeur du champ {{propertyName}} doit être entre {{valueMin}} et {{valueMax}}",
      "13": "La valeur du champ {{propertyName}} doit être numérique",
      "14": "La valeur du champ {{propertyName}} doit être égale à une de ces valeurs: {{valueMin}}, {{valueMax}}, ...",
      "30": "La valeur du champ {{propertyName}} doit être inférieure à {{valueMin}}",
      "31": "La valeur du champ {{propertyName}} doit être supérieure à {{valueMin}}",
      "32": "La valeur du champ {{propertyName}} doit être inférieure ou égale à {{valueMin}}",
      "33": "La valeur du champ {{propertyName}} doit être supérieure ou égale à {{valueMin}}",
      "34": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "35": "La valeur du champ {{propertyName}} doit être inclu entre {{valueMin}} et {{valueMax}}",
      "36": "La valeur du champ {{propertyName}} doit être inférieure à {{valueMin}}",
      "37": "La valeur du champ {{propertyName}} doit être supérieure à {{valueMin}}",
      "38": "La valeur du champ {{propertyName}} doit être inférieure ou égale à {{valueMin}}",
      "39": "La valeur du champ {{propertyName}} doit être supérieure ou égale à {{valueMin}}",
      "40": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "41": "La valeur du champ {{propertyName}} doit être égale à une de ces valeurs: {{valueMin}}, {{valueMax}}, ...",
      "60": "La valeur du champ {{propertyName}} ne peux pas être rempli",
      "61": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "62": "La valeur du champ {{propertyName}} doit être numérique",
      "63": "La longueur du champ {{propertyName}} n'est pas correcte ({{valueMin}})",
      "64": "La valeur du champ {{propertyName}} ne peux pas être rempli",
      "65": "La valeur du champ {{propertyName}} doit être égale à une de ces valeurs: {{valueMin}}, {{valueMax}}, ...",
      "66": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "67": "La valeur du champ {{propertyName}} doit avoir une longueur comprise entre {{valueMin}} et {{valueMax}}",
      "90": "Le champ {{propertyName}} est obligatoire",
      "91": "L'objet {{propertyName}} existe déjà dans l'application",
      "92": "L'objet {{propertyName}} n'a pas été trouvé",
      "93": "La valeur du champ {{propertyName}} est immutable et ne peux pas être modifié",
      "94": "La valeur du champ {{propertyName}} est obligatoire",
      "110": "Le nombre d'élements {{propertyName}} doit être compris entre {{valueMin}} et {{valueMax}}",
      "111": "Le nombre d'élements {{propertyName}} doit être compris entre {{valueMin}} et {{valueMax}}",
      "112": "La valeur du champ {{propertyName}} est en chevauchement avec {{valueMin}}",
      "113": "La valeur du champ {{propertyName}} ne se trouve pas dans la liste",
      "114": "L'objet {{propertyName}} n'est pas accessible par cet utilisateur",
      "130": "Le format du champ {{propertyName}} ne représente pas un numéro de téléphone valable",
      "131": "Le format du champ {{propertyName}} ne représente pas un e-mail valable",
      "150": "Le format du champ {{propertyName}} est incorrect",
      "199": "Un bug s'est produit"
    }
  },
  "SHARED": {
    "MSG": {
      "EMAIL_FORMAT": "L'adresse e-mail n'est pas valide.",
      "INVALID_IBAN": "L'IBAN n'est pas valide.",
      "BIC_FORMAT": "Le code BIC n'est pas valide."
    },
    "CALENDAR": {
      "TODAY": "Aujoudr'hui",
      "CLEAR": "Effacer"
    }
  },
  "MSG": "Message de test",
  "FORM": {
    "BANQUE": "Banque",
    "CONTACT": "Contact",
    "PERSONNE": "Personne",
    "IBAN": {
      "LABEL": "IBAN",
      "PLACEHOLDER": "Introduisez votre IBAN"
    },
    "BIC": {
      "LABEL": "BIC",
      "PLACEHOLDER": "Introduisez votre BIC"
    },
    "TELEPHONE": {
      "LABEL": "Téléphone",
      "PLACEHOLDER": "Introduisez votre n° de téléphone"
    },
    "GSM": {
      "LABEL": "GSM",
      "PLACEHOLDER": "Introduisez votre n° de GSM"
    },
    "FAX": {
      "LABEL": "Fax",
      "PLACEHOLDER": "Introduisez votre n° de fax"
    },
    "EMAIL": {
      "LABEL": "E-mail",
      "PLACEHOLDER": "Introduisez votre adresse e-mail"
    },
    "NOM": {
      "LABEL": "Nom",
      "PLACEHOLDER": "Introduisez votre nom"
    },
    "PRENOM": {
      "LABEL": "Prénom",
      "PLACEHOLDER": "Introduisez votre adresse prénom"
    },
    "NISS": {
      "LABEL": "N° National",
      "PLACEHOLDER": "Introduisez votre n° national"
    },
    "DATE_NAISSANCE": {
      "LABEL": "Date de naissance",
      "PLACEHOLDER": "Introduisez votre date de naissance"
    },
    "SEXE": {
      "LABEL": "Sexe",
      "PLACEHOLDER": "Sélectionnez votre sexe"
    },
    "LANGUE": {
      "LABEL": "Langue",
      "PLACEHOLDER": "Sélectionnez votre langue"
    }
  },
  "EXTERNAL": {
    "ERROR": {
      "201": "Le service {{source}} est indisponible : {{message}}"
    }
  }
}
```
Result using recursive merge:
```json
{
  "KERNEL": {
    "MSG": {
      "REQUIRED": "Ce champs est obligatoire.",
      "ERRORS_TITLE": "Erreur(s) bloquante(s)",
      "ERRORS_BADGE": "{{count}} erreur(s)",
      "WARNINGS_TITLE": "Alerte(s)",
      "WARNINGS_BADGE": "{{count}} alerte(s)",
      "INTERNAL_ERROR": "Une erreur s'est produite.",
      "DEFAULT_SUCCESS": "L'opération s'est déroulée correctement.",
      "DEFAULT_WARNING": "L'opération s'est déroulée avec des alertes.",
      "DEFAULT_ERROR": "Une erreur s'est produite."
    },
    "ERROR": {
      "1": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "2": "La valeur du champ {{propertyName}} doit être inférieure à {{valueMin}}",
      "3": "La valeur du champ {{propertyName}} doit être inférieure ou égale à {{valueMin}}",
      "4": "La valeur du champ {{propertyName}} doit être supérieure à {{valueMin}}",
      "5": "La valeur du champ {{propertyName}} doit être supérieure ou égale à {{valueMin}}",
      "6": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "7": "La valeur du champ {{propertyName}} doit être inférieure à {{valueMin}}",
      "8": "La valeur du champ {{propertyName}} doit être inférieure ou égale à {{valueMin}}",
      "9": "La valeur du champ {{propertyName}} doit être supérieure à {{valueMin}}",
      "10": "La valeur du champ {{propertyName}} doit être supérieure ou égale à {{valueMin}}",
      "11": "La valeur du champ {{propertyName}} doit être entre {{valueMin}} et {{valueMax}}",
      "13": "La valeur du champ {{propertyName}} doit être numérique",
      "14": "La valeur du champ {{propertyName}} doit être égale à une de ces valeurs: {{valueMin}}, {{valueMax}}, ...",
      "30": "La valeur du champ {{propertyName}} doit être inférieure à {{valueMin}}",
      "31": "La valeur du champ {{propertyName}} doit être supérieure à {{valueMin}}",
      "32": "La valeur du champ {{propertyName}} doit être inférieure ou égale à {{valueMin}}",
      "33": "La valeur du champ {{propertyName}} doit être supérieure ou égale à {{valueMin}}",
      "34": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "35": "La valeur du champ {{propertyName}} doit être inclu entre {{valueMin}} et {{valueMax}}",
      "36": "La valeur du champ {{propertyName}} doit être inférieure à {{valueMin}}",
      "37": "La valeur du champ {{propertyName}} doit être supérieure à {{valueMin}}",
      "38": "La valeur du champ {{propertyName}} doit être inférieure ou égale à {{valueMin}}",
      "39": "La valeur du champ {{propertyName}} doit être supérieure ou égale à {{valueMin}}",
      "40": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "41": "La valeur du champ {{propertyName}} doit être égale à une de ces valeurs: {{valueMin}}, {{valueMax}}, ...",
      "60": "La valeur du champ {{propertyName}} ne peux pas être rempli",
      "61": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "62": "La valeur du champ {{propertyName}} doit être numérique",
      "63": "La longueur du champ {{propertyName}} n'est pas correcte ({{valueMin}})",
      "64": "La valeur du champ {{propertyName}} ne peux pas être rempli",
      "65": "La valeur du champ {{propertyName}} doit être égale à une de ces valeurs: {{valueMin}}, {{valueMax}}, ...",
      "66": "La valeur du champ {{propertyName}} doit être égale à {{valueMin}}",
      "67": "La valeur du champ {{propertyName}} doit avoir une longueur comprise entre {{valueMin}} et {{valueMax}}",
      "90": "Le champ {{propertyName}} est obligatoire",
      "91": "L'objet {{propertyName}} existe déjà dans l'application",
      "92": "L'objet {{propertyName}} n'a pas été trouvé",
      "93": "La valeur du champ {{propertyName}} est immutable et ne peux pas être modifié",
      "94": "La valeur du champ {{propertyName}} est obligatoire",
      "110": "Le nombre d'élements {{propertyName}} doit être compris entre {{valueMin}} et {{valueMax}}",
      "111": "Le nombre d'élements {{propertyName}} doit être compris entre {{valueMin}} et {{valueMax}}",
      "112": "La valeur du champ {{propertyName}} est en chevauchement avec {{valueMin}}",
      "113": "La valeur du champ {{propertyName}} ne se trouve pas dans la liste",
      "114": "L'objet {{propertyName}} n'est pas accessible par cet utilisateur",
      "130": "Le format du champ {{propertyName}} ne représente pas un numéro de téléphone valable",
      "131": "Le format du champ {{propertyName}} ne représente pas un e-mail valable",
      "150": "Le format du champ {{propertyName}} est incorrect",
      "199": "Un bug s'est produit"
    }
  },
  "SHARED": {
    "MSG": {
      "EMAIL_FORMAT": "L'adresse e-mail n'est pas valide.",
      "INVALID_IBAN": "L'IBAN n'est pas valide.",
      "BIC_FORMAT": "Le code BIC n'est pas valide."
    },
    "CALENDAR": {
      "TODAY": "Aujoudr'hui",
      "CLEAR": "Effacer"
    }
  },
  "MSG": "Message de test",
  "FORM": {
    "BANQUE": "Banque",
    "CONTACT": "Contact",
    "PERSONNE": "Personne",
    "IBAN": {
      "LABEL": "IBAN",
      "PLACEHOLDER": "Introduisez votre IBAN"
    },
    "BIC": {
      "LABEL": "BIC",
      "PLACEHOLDER": "Introduisez votre BIC"
    },
    "TELEPHONE": {
      "LABEL": "Téléphone",
      "PLACEHOLDER": "Introduisez votre n° de téléphone"
    },
    "GSM": {
      "LABEL": "GSM",
      "PLACEHOLDER": "Introduisez votre n° de GSM"
    },
    "FAX": {
      "LABEL": "Fax",
      "PLACEHOLDER": "Introduisez votre n° de fax"
    },
    "EMAIL": {
      "LABEL": "E-mail",
      "PLACEHOLDER": "Introduisez votre adresse e-mail"
    },
    "NOM": {
      "LABEL": "Nom",
      "PLACEHOLDER": "Introduisez votre nom"
    },
    "PRENOM": {
      "LABEL": "Prénom",
      "PLACEHOLDER": "Introduisez votre adresse prénom"
    },
    "NISS": {
      "LABEL": "N° National",
      "PLACEHOLDER": "Introduisez votre n° national"
    },
    "DATE_NAISSANCE": {
      "LABEL": "Date de naissance",
      "PLACEHOLDER": "Introduisez votre date de naissance"
    },
    "SEXE": {
      "LABEL": "Sexe",
      "PLACEHOLDER": "Sélectionnez votre sexe"
    },
    "LANGUE": {
      "LABEL": "Langue",
      "PLACEHOLDER": "Sélectionnez votre langue"
    }
  },
  "EXTERNAL": {
    "ERROR": {
      "201": "Le service {{source}} est indisponible : {{message}}"
    }
  }
}
```
The main difference is visible in the KERNEL key.
For creating the example of recursive merge, I used the `lodash` method `_.merge`. I know you might want to not add a dependency to lodash but you might want to check how they do it.

**Specifications: Angular version, library version, environment, browser**
Angular 5.03
angular-l10n 4.0.0

**Comment by robisim74** [comment]
@FabienDehopre This is the expected behavior for translation data: duplicate keys must be overwritten (in each level), and the _spread_ operator was chosen specifically. In any case I would never use lodash as a library dependency.
This allows mechanisms such as the fallback providers to work (that you can find here: https://robisim74.github.io/angular-l10n/spec/configuration/#loading-the-translation-data)

**Comment by FabienDehopre** [comment]
I understand that you want that duplicate keys must be overwritten but the spread operator only operates on the first level of keys. A recursive approach would do the same but would allow to have deeper composite keys as in my example.
I also understand that you do not want to use lodash but it was just for the example.
Would you mind if I try to do a PR for implementing a recursive merge as described here?

**Comment by robisim74** [comment]
PRs are always welcome, especially if they propose a better solution.

**Comment by robisim74** [comment]
@FabienDehopre Apparently my fears about the performance of deep merging were not founded: I measured the performance in a base scenario (a few hundred keys in 4 files) and then with about 45000 and 90000 keys. There seems to be no difference, both methods rise linearly, and remain in a delta of about 10ms.
So I adopted deep merging by default. The new version is out: https://github.com/robisim74/angular-l10n/releases

Thanks so much!

Greetings,
Roberto
